### PR TITLE
move hidden fields of leaftype datatypes into a separate struct

### PR DESCRIPTION
**Pull request by vtjnash**
this shaves a sizeable amount off the system image (almost 10%)

### CODE REVIEW

**Comment by carnaval** [comment]
Cool !
As far as I can tell there is no hash consing in there yet, so is the saving only coming from a few bytes on every non-leaf type ?
I guess starting from that it should be pretty easy to just check how much duplication there is in there and if the uniquing is worth it.

File: src/dump.c [hunk 1]
```diff
@@ -692,6 +692,8 @@ static int literal_val_id(jl_value_t *v)
 
 static void jl_serialize_value_(ios_t *s, jl_value_t *v)
 {
+    //write_uint16(s, 0xABCD);
```
Comment (JeffBezanson) [inline src/dump.c]: ?

File: src/julia.h [hunk 2]
```diff
@@ -304,55 +304,58 @@ typedef struct {
...
+    const struct _jl_datatype_layout_t *layout;
+    int32_t size; // TODO: move to _jl_datatype_layout_t
+    int32_t ninitialized;
+    uint32_t uid;
     uint8_t abstract;
     uint8_t mutabl;
-    uint8_t pointerfree;
-    int32_t ninitialized;
     // memoized properties
+    void *struct_decl;  //llvm::Type*
```
Comment (JeffBezanson) [inline src/julia.h]: Maybe move these to the end to avoid empty symbol field names?

File: base/inference.jl [hunk 3]
```diff
@@ -2322,7 +2322,7 @@ function inlineable(f::ANY, ft::ANY, e::Expr, atypes::Vector{Any}, sv::Inference
     end
     if length(atypes)==3 && is(f,unbox)
         at3 = widenconst(atypes[3])
-        if isa(at3,DataType) && !at3.mutable && at3.pointerfree
+        if isa(at3,DataType) && !at3.mutable && at3.layout != C_NULL && datatype_pointerfree(at3)
```
Comment (JeffBezanson) [inline base/inference.jl]: I would put the `layout != C_NULL` check inside the `datatype_pointerfree` function to make it more robust.

Comment (tkelman) [inline base/inference.jl]: why wasn't this or the empty field names suggestion addressed?

File: src/dump.c [hunk 4]
```diff
@@ -555,35 +555,35 @@ static void jl_serialize_datatype(ios_t *s, jl_datatype_t *dt)
         return;
     }
 
-    size_t nf = jl_datatype_nfields(dt);
-    write_uint16(s, nf);
     write_int32(s, dt->size);
     int has_instance = !!(dt->instance != NULL);
-    write_uint8(s, dt->abstract | (dt->mutabl<<1) | (dt->pointerfree<<2) | (has_instance<<3) |
+    int has_layout = !!(dt->layout != NULL);
```
Comment (tkelman) [inline src/dump.c]: why the `!!` ?

File: base/inference.jl (hunk 3 above)
Comment (vtjnash) [inline base/inference.jl]: both were fixed

File: src/dump.c (hunk 4 above)
Comment (vtjnash) [inline src/dump.c]: it's a common idiom for turning a number into a 0 or 1 (http://stackoverflow.com/questions/10307281/c-operator-is-a-two-not)

Comment (yuyichao) [inline src/dump.c]: it's redundant with the `!= NULL` though...

**Comment by tkelman** [comment]
This broke JLD, cc @timholy @simonster

**Pull request merged**